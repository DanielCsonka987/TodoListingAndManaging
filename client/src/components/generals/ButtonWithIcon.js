import React, { useState } from 'react'

function ButtonWithIcon(props){
    let btnConfig = [];
    let btnTextContent = ''
    const [ tooltipBubbleState, setTooltipState ] = useState(false)
    
    if(props.sizing=== 'big'){
        btnConfig[0] = 'btnObjBig ';
        btnConfig[1] = 'iconAndTextTogether '
        if(props.children){
            btnTextContent =  <span className={'textNextToIcon ' + props.textClasses}>
                 {props.children} </span>
        }
    }else{  //at small mode, the button text appears as bubble at focus/noMouse
        btnConfig[0] = 'btnObjSmall '
        btnConfig[1] = 'iconWithoutText '
        if(props.children){
            btnTextContent =  tooltipBubbleState ? 
                <span className={'btnTooltipBubble'} name={props.naming} >
                    {  props.children } </span>  : ''
        }
    }
    const bubbleAppeare = ()=>{
        setTooltipState(true)
    }
    const bubbleDisappeare = ()=>{
        setTooltipState(false)
    }
    
    return(
        <button name={props.naming} 
            className={'clickable ' + btnConfig[0] + props.wrapperBlockClasses }
            onClick={props.funcClickActivity}
            onFocus={bubbleAppeare} onMouseOver={bubbleAppeare}
            onBlur={bubbleDisappeare} onMouseOut={bubbleDisappeare}>
            <span name={props.naming} className={btnConfig[1] + props.wrapperInlineClasses } >
                <i name={props.naming} className='material-icons iconItself' >
                    {props.iconDef}</i>
                { btnTextContent }
            </span>
        </button>
    )
}

export default ButtonWithIcon