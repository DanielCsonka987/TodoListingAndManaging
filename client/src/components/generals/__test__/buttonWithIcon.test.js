import React from 'react'
import { waitFor, fireEvent, screen, render } from '@testing-library/react'
import ButtonWithIcon from '../ButtonWithIcon'
import '@testing-library/jest-dom/extend-expect'
import renderer from 'react-test-renderer';
import { toMatchDiffSnapshot } from 'snapshot-diff'
expect.extend({ toMatchDiffSnapshot })

describe('Button complex, big and small version', ()=>{

    test('Component test - big config structure and activity', async ()=>{

        const testingFunc = jest.fn()
        const comp = <ButtonWithIcon 
            sizing='big' wrapperBlockClasses='noticable'
            naming='buttonAndInner'
            iconDef='cancel' funcClickActivity={ testingFunc }
        
        >Cancel button</ButtonWithIcon>
        const tree = renderer.create(comp).toJSON()
        expect(tree).toMatchSnapshot()
        
        const { container, unmount } = render(comp)
        const wrapperContainer = container.firstChild
        expect(wrapperContainer).toHaveClass('noticable')
    
        const theButton = screen.getByRole('button')
        expect(theButton).toBeInTheDocument()
        expect(theButton).toHaveClass('clickable')
        expect(theButton).toHaveClass('btnObjBig')
        expect(theButton).toHaveClass('noticable')
        expect(theButton).toHaveAttribute('name', 'buttonAndInner')
    
        const buttonContent = theButton.firstChild
        expect(buttonContent).toHaveClass('iconAndTextTogether')
        expect(buttonContent).toHaveAttribute('name', 'buttonAndInner')
    
        const iconArea = screen.getByText('cancel')
        expect(iconArea).toHaveClass('material-icons')
        expect(iconArea).toHaveClass('iconItself')
        expect(iconArea).toHaveAttribute('name', 'buttonAndInner')
        const textArea = screen.getByText('Cancel button')
        expect(textArea).toHaveClass('textNextToIcon')
    
        expect(buttonContent).toContainElement(iconArea)
        expect(buttonContent).toContainElement(textArea )
        
        fireEvent.click(theButton)
        await waitFor(()=>{
            expect(testingFunc).toHaveBeenCalled()
        })
        unmount()
        
    })

    test('Component test - small config structure and activity', async ()=>{
        let testingFunc = jest.fn()
        const comp = <ButtonWithIcon 
            classes = 'identifier'
            sizing='small' wrapperBlockClasses='noticable'
            naming='buttonAndInner'
            iconDef='cancel' funcClickActivity={ testingFunc }
        
        >Cancel button</ButtonWithIcon>
        
        const { act } = renderer
        const theComponent = renderer.create(comp)
        const tree = theComponent.toJSON()
        expect(tree).toMatchSnapshot()
        
        const { container, unmount, debug } = render(comp)
        const theButton = screen.getByRole('button')
        expect(theButton).toBeInTheDocument()
        expect(theButton).toHaveClass('clickable')
        expect(theButton).toHaveClass('btnObjSmall')
        expect(theButton).toHaveClass('noticable')
        expect(theButton).toHaveAttribute('name', 'buttonAndInner')

        const iconArea = screen.getByText('cancel')
        expect(iconArea).toHaveClass('material-icons')
        expect(iconArea).toHaveClass('iconItself')
        expect(iconArea).toHaveAttribute('name', 'buttonAndInner')
        
        expect(screen.queryByText('Cancel button')).not.toBeInTheDocument()

        fireEvent.focusIn(theButton)    //popup appeares
        await waitFor( ()=>{
            const textArea = screen.getByText('Cancel button')
            expect(textArea).toBeInTheDocument()
            
            act(()=>{
                theComponent.root.findAllByType('button')[0].props.onFocus()
            })
            const treeUpdate = theComponent.toJSON()
            expect(tree).toMatchDiffSnapshot(treeUpdate)
        })
        
        fireEvent.focusOut(theButton)
        await waitFor(()=>{
            expect(screen.queryByText('Cancel button')).not.toBeInTheDocument()
        })
        fireEvent.click(theButton)
        await waitFor(()=>{
            expect(testingFunc).toHaveBeenCalled()
        })
        unmount()
    })

})