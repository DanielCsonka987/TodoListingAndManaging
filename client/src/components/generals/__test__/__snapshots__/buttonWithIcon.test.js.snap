// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button complex, big and small version Component test - big config structure and activity 1`] = `
<button
  className="clickable btnObjBig noticable"
  name="buttonAndInner"
  onBlur={[Function]}
  onClick={[MockFunction]}
  onFocus={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
>
  <span
    className="iconAndTextTogether undefined"
    name="buttonAndInner"
  >
    <i
      className="material-icons iconItself"
      name="buttonAndInner"
    >
      cancel
    </i>
    <span
      className="textNextToIcon undefined"
    >
      Cancel button
       
    </span>
  </span>
</button>
`;

exports[`Button complex, big and small version Component test - small config structure and activity 1`] = `
<button
  className="clickable btnObjSmall noticable"
  name="buttonAndInner"
  onBlur={[Function]}
  onClick={[MockFunction]}
  onFocus={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
>
  <span
    className="iconWithoutText undefined"
    name="buttonAndInner"
  >
    <i
      className="material-icons iconItself"
      name="buttonAndInner"
    >
      cancel
    </i>
    
  </span>
</button>
`;

exports[`Button complex, big and small version Component test - small config structure and activity 2`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -15,8 +15,14 @@
        className=\\"material-icons iconItself\\"
        name=\\"buttonAndInner\\"
      >
        cancel
      </i>
-     
+     <span
+       className=\\"btnTooltipBubble\\"
+       name=\\"buttonAndInner\\"
+     >
+       Cancel button
+        
+     </span>
    </span>
  </button>"
`;
